### ISSUE TITLE
aarch64: test TestMountMoreThan42LayersMatchingPathLength fails (aufs)

### ISSUE THREAD

**Issue by vielmetti** [moby/moby#27357]
**Description**

On aarch64, the test TestMountMoreThan42LayersMatchingPathLength fails.

**Steps to reproduce the issue:**
1. Build Docker on aarch64 according to http://blog.hypriot.com/post/getting-docker-running-on-a-high-density-armv8-server-from-hisilicon/
2. Using the 1.10.2 image you built, build docker in docker for 1.13.0-dev
3. Get to "Making bundle: test-unit (in bundles/1.13.0-dev/test-unit)"
4. Fail the test; details of the fail at https://gist.github.com/33be26e0b7e634e01ad12ea672e39472

**Describe the results you received:**

aufs tests fail with error

time="2016-10-13T15:33:39Z" level=error msg="Can't add file /tmp/aufs-tests/aufs/diff/1/test_file to tar: lstat /tmp/aufs-tests/aufs/diff/1/test_file: no such file or directory" 

the test TestMountMoreThan42LayersMatchingPathLength fails.

**Describe the results you expected:**

aufs tests pass

**Additional information you deem important (e.g. issue happens only occasionally):**

This is my first time doing a full build from source on this platform, and I was not able to run check-config.sh to know that the kernel support for aufs is complete. Unclear if bug in test or in filesystem.

**Output of `docker version`:**

root@armv8hello:~/src/docker# docker version
Client:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.6.2
 Git commit:   c3959b1
 Built:        Thu Oct 13 14:55:27 2016
 OS/Arch:      linux/arm64

Server:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.6.2
 Git commit:   c3959b1
 Built:        Thu Oct 13 14:55:27 2016
 OS/Arch:      linux/arm64

**Output of `docker info`:**

root@armv8hello:~/src/docker# docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 33
Server Version: 1.10.2
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 25
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: null host bridge
Kernel Version: 4.4.0-38-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: aarch64
CPUs: 96
Total Memory: 125.8 GiB
Name: armv8hello.local.lan
ID: QB72:5PAI:O62E:QEPU:4DKN:3M2K:33FC:AIJU:PXMC:YMGX:Y376:IAPK
WARNING: No swap limit support

**Additional environment details (AWS, VirtualBox, physical, etc.):**

ARMv8 96 core server under test.

**Comment by justincormack** [comment]
There were a bunch of failing tests on aarch64 last time I did a test build. There is a bit of work needing doing on this port. This particular issue though seems more likely to be a kernel issue. I would actually recommend using `overlay2` as your kernel is recent.

**Comment by vielmetti** [comment]
Thanks for the suggestion @justincormack. The consensus from the vendor is that 1.12.2 (instead of 1.13.0-dev) is a good next target, so I'm aiming at that. One snag was a dependency on `containerd` that I don't completely understand.

**Comment by justincormack** [comment]
containerd is a dependency for both. If you want to fix things, you need to submit against master; I doubt that the differences and fixes will be much different for the issues on aarch64. There are a few open and closed issues that you could take a look at.

**Comment by docbobo** [comment]
Is it possible that this goes away with Go 1.8 as well? I was running `make test-unit` with a Go 1.8 based version of the Dockerfiles and didn't run into this issue anymore.

**Comment by justincormack** [comment]
It may well be fixed in 1.8 at least if your machine has 4k pages. On 64k
pages eg Centos not sure...

**Comment by docbobo** [comment]
Well, I was running the test on packet.net with the type 2a machines. So basically the same setup as in the initial report. What would be a good way to verify the page size?

**Comment by vielmetti** [comment]
@docbobo - you can use code from https://github.com/vielmetti/go-pagesize-test to determine and verify page size

**Comment by vielmetti** [comment]
Just confirming that this bug still appears when I build Docker on Ubuntu 16.04 on Packet 2A with this setup:

```
root@docker-build-test:~# docker version
Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Tue Mar 21 17:07:57 2017
 OS/Arch:      linux/arm64

Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Tue Mar 21 17:07:57 2017
 OS/Arch:      linux/arm64
 Experimental: false
```

**Comment by vielmetti** [comment]
OK, I think I found the bug.

Look at 

https://github.com/docker/docker/blob/2c2983cd8e58630889712c40a614714ce0c68efc/daemon/graphdriver/aufs/aufs.go#L577

where the call is  

```
        // Mount options are clipped to page size(4096 bytes). If there are more
	// layers then these are remounted individually using append.

	offset := 54
	if useDirperm() {
		offset += len("dirperm1")
	}
	b := make([]byte, syscall.Getpagesize()-len(mountLabel)-offset) // room for xino & mountLabel
	bp := copy(b, fmt.Sprintf("br:%s=rw", rw))
```

However, page size is not always 4096 bytes. On ARMv8 (aarch64) machines, it can vary, depending on the operating system. To make matters worse, Go < 1.8 returns the a too-big page size for some machines.

The overlay2 filesystem had the same bug, and here's how it was addressed:

https://github.com/albamc/docker/commit/215ee2986f290b01e3f337eedbfc9a4a62f26c81

with a note that the issue would be moot when Go 1.8 is supported.
